### ISSUE TITLE
[Bug]: JUCE headers can fail guarding themselves against macros defined by windows.h

### ISSUE THREAD

**Issue by danra** [juce-framework/JUCE#1298]
### Detailed steps on how to reproduce the bug

1) Start an example GUI application on Windows
2) Add the following to MainComponent.h:

```
#include <juce_core/juce_core.h>
#include <windows.h>
#include <juce_gui_basics/juce_gui_basics.h>
```

Get 20 errors (tested in VS2022) due to the `min` and `max` macros not being undefined properly (which JUCE does try to do, both by using `NOMINMAX` and by `#undef`ing them in a few places).
Replacing `juce_gui_basics` with `juce_gui_extra` gives even more errors due to the standard `small` macro being used as an enum value in `juce_PushNotifications.h`.

My attempted analysis at why JUCE's defenses aren't working properly:
- `juce_core.h` is first included. `JUCE_CORE_INCLUDE_NATIVE_HEADERS` is not defined, so `native/juce_BasicNativeHeaders.h` isn't included and so neither is `windows.h`. `juce_core.h` then undefines the standard `small` macro on its own and some more macros in `juce_StandardHeader.h`, but since `windows.h` wasn't included these aren't defined anyway.
- `windows.h` is then included without the useful `WIN32_LEAN_AND_MEAN`, `NOMINMAX` etc.
- `juce_gui_basics.h` is then included, but `juce_core.h` of course isn't included again because of the `#pragma once`. So none of the `#undef`s happen, and the standard Windows macros interfere with the JUCE definitions.

The above example is of course artificial, but we did actually get this in a more realistic scenario (also simplified here):

`foo.cpp`:
```
#include "foo.h"

#include "bar.h"

#include <juce_gui_extra/juce_gui_extra.h>
...
```

Header `bar.h`:
```
#pragma once

#include <juce_audio_processors/juce_audio_processors.h>

#include <third_party.h>
...
```

Header `third_party.h`:
```
...
#include <windows.h>
...
```

### What is the expected behaviour?

Including a juce module, followed by including `windows.h`, followed by including another juce module should always work.

### Operating systems

Windows

### What versions of the operating systems?

all

### Architectures

x86_64

### Testing on the `develop` branch

The bug is present on the `develop` branch

### Code of Conduct

- [X] I agree to follow the Code of Conduct

**Comment by mkruselj** [comment]
We also have this happening in Shortcircuit XT, constant warnings about NOMINMAX macro redefinition. We traced it to juce_BasicNativeHeaders.h not having a guard around that `#define NOMINMAX` at around line 80-something.

Would be nice if this were guarded 🙂

**Comment by mkruselj** [comment]
@reuk Any chance for this small buglet to be fixed any time soon? 😅

**Comment by Anthony-Nicholls** [comment]
I don't feel completely comfortable with this. It feels wrong for JUCE to define something like `NOMINMAX` or `WIN32_LEAN_AND_MEAN` when JUCE isn't the one including `Windows.h`. For all JUCE knows a 3rd party library might need those definitions to remain undefined. 

It also feels brittle for JUCE to undefine the problematic preprocessor definitions. This would either need to be done before the `#pragma once` in `juce_core.h` (assuming every module does indeed depend on `juce_core`) or in each juce module header independently. 

Although min/max/small might be common ones that occur becuase of Windows.h, anything could occur from any third party header. I'm not sure it's JUCE's job to guard against this unless JUCE itself is the one including the header.

Some alternative solutions...

Create your own header
- Create a header that defines the definitions you want, then includes the offending header
- Create a header that includes the offending header, then undefines the problematic preprocessor definitions

Wrap up the functionality into a proxy or pimpl object
Basically only include the offending header from a cpp file that contains the implementation of functions / classes that you have defined in another header that doesn't include Windows.h.

Alternatively contact the authors of the third party library and ask them to implement one of the above.

**Comment by mkruselj** [comment]
Exactly!

**Comment by mkruselj** [comment]
I would need to check in Shortcircuit XT code, it could be that one of 3rd party libraries is declaring it and then you can imagine the house of cards tumbling down...

**Comment by baconpaul** [comment]
We don’t use the juceaide build but use modules only approach which may be why it gets included

But in any case, guarding it seems like best practice for all the reasons given

**Comment by mkruselj** [comment]
More than half a year later... it seems like this issue ended up a bit forgotten, yeah? Seems like a very simple thing to do... just guard your defines...
